### PR TITLE
[C++] Optimize ATN deserialization

### PR DESCRIPTION
**Pull request by jcking**
Improve ATN deserialization by avoiding unnecessary copies and right-sizing `std::vector` when the size is known ahead of time.

### CODE REVIEW

**Comment by jcking** [comment]
@mike-lischke

File: runtime/Cpp/runtime/src/atn/ATNDeserializer.cpp [hunk 1]
```diff
@@ -39,6 +39,7 @@
 #include "Exceptions.h"
 #include "support/CPPUtils.h"
 #include "support/StringUtils.h"
+#include "support/Casts.h"
```
Comment (mike-lischke) [inline runtime/Cpp/runtime/src/atn/ATNDeserializer.cpp:42]: Seems this file is missing.

**Review by mike-lischke — COMMENTED** [review]
I like the clean up of the code (casts in particular), but I don't really see the advantage of use the `SerializedATNView` class. Does it really speed up deserialization? Do you have any evidence?

**Comment by jcking** [comment]
`SerializedATNView` avoids having to make yet another copy of the serialized ATN. This is already done once during static initialization (before `main`), and this would require it be done again resulting in 2 copies of the serialized ATN. Allocations are slow and avoiding them improves speed, especially when we really only need to subtract 2 for each element. The `SerializedATNView` basically just wraps the vector and subtracts 2 at each access, to keep code simple.

**Review by mike-lischke — APPROVED** [review]
Increase complexity to make things smaller and faster. Yeah, that's what C++ is known for 🤣

**Comment by mike-lischke** [comment]
@parrt OK, this one is also ready for merge.

**Pull request merged**